Content.Shared/Storage/SharedItemCounterComponent.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.Manager.Attributes;
+
+namespace Content.Shared.Storage
+{
+    [RegisterComponent]
+    public class SharedItemCounterComponent : Component, ISerializationHooks
+    {
+        public override string Name => "ItemCounter";
+
+        [DataField("mapLayers")] public readonly List<LayerProperties> _mapLayers = new();
+        public IReadOnlyList<string> SpriteLayers = new List<string>();
+
+        [Serializable]
+        [DataDefinition]
+        public struct LayerProperties
+        {
+            [DataField("layer")] public string Layer;
+            [DataField("ids")] public List<string>? Id { get; set; }
+            [DataField("tags")] public List<string>? Tags { get; set; }
```

Comment:
see if you can use an EntityWhitelist for this (recently added), it basically is used for this exact purpose, and you can remove the ugly Matches extension method on entitites. it doesn't have support for prototype IDs (since its usually not a good idea) but it can be added if necessary, or you can just use the tag support it has